src/metabase/driver/sql/query_processor.clj

```diff
@@ -56,6 +57,23 @@
          (= sql    (.sql ^SQLSourceQuery other))
          (= params (.params ^SQLSourceQuery other)))))
 
+(alter-meta! #'->SQLSourceQuery assoc :private true)
+
+(defn sql-source-query
+  "Preferred way to construct an instance of [[SQLSourceQuery]]. Does additional validation."
+  [sql params]
+  (when-not (string? sql)
+    (throw (ex-info (tru "Expected native source query to be a string, got: {0}"
+                         (.getCanonicalName (class sql)))
+                    {:type  qp.error-type/invalid-query
+                     :query sql})))
+  (when-not ((some-fn nil? sequential?) params)
+    (throw (ex-info (tru "Expected native source query parameters to be sequential, got: {0}"
+                         (.getCanonicalName (class params)))
+                    {:type  qp.error-type/invalid-query
+                     :query params})))
+  (SQLSourceQuery. sql params))
```

Comment:
This is not actually related to the test, but when writing the test I wrote it wrong and my query blew up with a pretty opaque error message. It turned out that I had accidentally passed in `{:native {:native "SELECT ..."}}` instead of `{:name "SELECT ...}}`. So this stuff here is just to give us better error messages for next time around